### ISSUE TITLE
HTML tags visible in headers on some Wikivoyage pages

### ISSUE THREAD

**Issue by Jaifroid** [openzim/mwoffliner#178]
@kelson42 : For example in `wikivoyage_en_all_novid_2017-12.zim`, visit the page entitled "Hot Springs". The HTML extracted from this ZIM looks like this on many of the headers:

`<span class="mw-headline" id="Africa }}">&lt;a href="/wiki/Africa" title="Africa"&gt;Africa&lt;/a&gt;</span>`

... resulting in a rendered view like the screenshot below. I don't know if the problem is caused by the `}}` in the id, in the original source of the page, or in mwoffliner's scraping process. I've seen something like this on a lot of Wikivoyage pages from the December ZIM, so I'm guessing it's a new issue or it would have been picked up before.

![image](https://user-images.githubusercontent.com/4304337/34072030-23214834-e278-11e7-924d-7618ce98fc96.png)

**Comment by kelson42** [comment]
@Jaifroid I can not confirm the problem. The HTML looks good on my end

```
$ zimdump -f "Hot_Springs" -d wikivoyage/wikivoyage_en_all_novid_2017-12.zim | grep "mw-headline"
    <span class="mw-headline" id="title_0">Hot Springs</span>
```

**Comment by Jaifroid** [comment]
@kelson42 Sorry, I wasn't clear enough -- the issue isn't in the title of Hot Springs, it's in the headings further down in the body of that article -- see the Africa heading half-way down the Hot Springs article. In the raw HTML you should see that we've got some `&lt;` and `&gt;` codes erroneously in the headings.

**Comment by Jaifroid** [comment]
@kelson42 NB article is "Hot springs" not "Hot Springs" (case difference!!)

**Comment by kelson42** [comment]
@Jaifroid I think I understood correctly. Here is the whole HTML as stored in the ZIM file. If you do not get the same content then something is wrong somewhere. [Hot_Springs.html.txt](https://github.com/openzim/mwoffliner/files/1565512/Hot_Springs.html.txt)

**Comment by Jaifroid** [comment]
That's the wrong article @kelson42 , it's lowercase s, so "Hot springs" or Hot_springs...

**Comment by Jaifroid** [comment]
Whole html of article should look like this: 
[Hot_springs.txt](https://github.com/openzim/mwoffliner/files/1565534/Hot_springs.txt)

**Comment by kelson42** [comment]
@Jaifroid I confirm the problem. Will have a look.

**Comment by bradyhunsaker** [comment]
Just documenting the issue and scope.  I'll start exploring the source code for a fix, though the code is new to me.

This happens when there is a heading that is also a link.  It affects both internal and external links.
Like: ===[[Australia]]===

It appears to not affect Wikipedia, however, despite the presence of the same markup.  Example:
zimdump -f "List_of_calculus_topics" -d wikipedia_en_maths_nopic_2018-06.zim | grep mw-headline
(see the Differential calculus heading)

Impact, based on searching in a few languages
- Wikivoyage lists 1407 (en), 583 (de), 114 (fr) pages with internal links; only about 100 with external links in those languages

Example search link:  https://en.wikivoyage.org/w/index.php?search=insource%3A%2F%3D%3D%5C%5B%5C%5B%2F&title=Special:Search&profile=default&fulltext=1

Another example Wikivoyage page showing the undesired handling:
zimdump -f "UNESCO_World_Heritage_List" -d wikivoyage_en_all_nopic_2018-06.zim | grep mw-headline

**Comment by bradyhunsaker** [comment]
When I try to replicate locally using mostly default settings, I get quite different text that does not have this problem.  I see <section> tags, so I believe it is using Parsoid v 0.9.0.  In contrast, the official builds do not have <section> tags, so I believe they are generated with an earlier version of Parsoid.

Is the setup to generate the official downloads somewhere in the repositories for me to try to recreate?

**Comment by kelson42** [comment]
@bradyhunsaker Thx for trying to fix that, I'll have a look again. Give me a few days.

**Comment by kelson42** [comment]
@bradyhunsaker The code to scrappe wikivoyage in English is here https://github.com/kiwix/maintenance/blob/master/mwoffliner/mwoffliner3_dump.sh#L157

Version of Parsoid used is parsoid@^0.8.0.

But it seems to be the correct version?
* https://github.com/openzim/mwoffliner/blob/master/package-lock.json#L1411
* https://github.com/openzim/mwoffliner/blob/master/package.json#L48

@bradyhunsaker Do you mean we should just upgrade? (please go ahead with a patch then, would be happy to confirm this fix the pb).

**Comment by bradyhunsaker** [comment]
Thanks for the pointer to the scraping code.

** I noticed that my comment above is missing an important piece of evidence, because I didn't realize HTML would be interpreted.  
My local test has \<section\> tags, which were added to Parsoid 0.9.0 according to the changelog.  That's my reason for believing that I was using Parsoid 0.9.0.  I do not see \<section\> tags in production Wikipedia, so my guess is that Parsoid 0.9.0 is not used there either.

I'm just learning, but my sense is that sticking with the same Parsoid version that production Wikipedia uses is a good idea, and of course I wouldn't upgrade without a bunch of testing that other things aren't broken.

I'll see if I can use the pointers you gave to recreate some production pages and identify the issue (even though it's possible that an eventual upgrade to 0.9.0 will also remove it).

**Comment by subbuss** [comment]
Note that Parsoid is currently not used for desktop read views on wikipedias yet. It is used for visual editing, and mobile apps which use the section tags added by Parsoid. The Parsoid version deployed there is always the latest entry on https://www.mediawiki.org/wiki/Parsoid/Deployments which is much further along than 0.9.0 (which was released in April).

**Comment by bradyhunsaker** [comment]
I just submitted PR #348 to fix this.

Thanks for the explanation, subbuss.  I hadn't realized the two parser issue, and now I read a couple of pages describing the situation and goals.

@kelson42 Your code pointer helped me recreate the problem, which was key.

It turned out the difference I was seeing was do to the different handling for mobile mode and desktop mode in mwoffliner.  It's not a parser issue.  Wikivoyage is done in mobile mode, and one template file was still autoescaping titles.

**Comment by kelson42** [comment]
@subbuss Thx for the feedback, the version numbering on npmjs is bit strange https://www.npmjs.com/package/parsoid. Should not be 0.9.1 the latest version?

**Comment by subbuss** [comment]
We haven't yet released 0.9.1 .. we meant to do it couple weeks back, but so much else going on that we couldn't get to it.
